File: packages/cli-python/mcpseal/mcp_client.py
# Mirrors packages/cli-node/src/mcp-client.ts: a minimal MCP stdio client,
# newline-delimited JSON-RPC (empirically confirmed against a real reference
# server — see docs/Tasks.md 2.1). Lives in the CLI package, not mcpseal's
# hashing/drift core, for the same reason as the TS version: spawning
# processes and speaking a wire protocol is CLI plumbing, not the
# canonical, trust-critical hashing spec.
from __future__ import annotations

import json
import subprocess
import threading
import time
from typing import Any, TypedDict

from mcpseal.process_utils import USE_SHELL, kill_process_tree
from mcpseal.version import VERSION

DEFAULT_TIMEOUT_S = 15.0


class McpToolDefinition(TypedDict, total=False):
    name: str
    description: str
    inputSchema: dict


class McpStdioClient:
    def __init__(self, command: str, args: list[str], cwd: str | None = None):
        self._proc = subprocess.Popen(
            [command, *args],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            # Forward the spawned process's stderr so diagnostics (e.g. a
            # blocked-tool log line) are visible, not silently swallowed —
            # matches the TS client piping child.stderr to process.stderr.
            stderr=None,
            cwd=cwd,
            shell=USE_SHELL,
            text=True,
            bufsize=1,
        )
        self._next_id = 1
        self._id_lock = threading.Lock()
        self._cond = threading.Condition()
        self._responses: dict[int, dict] = {}
        self._closed = False
        self._close_error: Exception | None = None
        self._reader = threading.Thread(target=self._read_loop, daemon=True)
        self._reader.start()

    def _read_loop(self) -> None:
        try:
            assert self._proc.stdout is not None
            for raw_line in self._proc.stdout:
                line = raw_line.rstrip("\n")
                if not line.strip():
                    continue
                try:
                    msg = json.loads(line)
                except json.JSONDecodeError:
                    continue  # not a JSON-RPC message we can correlate; ignore
                if not isinstance(msg, dict) or msg.get("id") is None:
                    continue  # notification, not a response to a request
                with self._cond:
                    self._responses[msg["id"]] = msg
                    self._cond.notify_all()
        except Exception:
            pass
        finally:
            with self._cond:
                self._closed = True
                if self._close_error is None:
                    self._close_error = RuntimeError("server process exited before responding")
                self._cond.notify_all()

    def _request(self, method: str, params: Any, timeout_s: float = DEFAULT_TIMEOUT_S) -> Any:
        if self._closed:
            raise self._close_error or RuntimeError("MCP client is closed")
        with self._id_lock:
            req_id = self._next_id
            self._next_id += 1
        payload = json.dumps({"jsonrpc": "2.0", "id": req_id, "method": method, "params": params})
        assert self._proc.stdin is not None
        self._proc.stdin.write(payload + "\n")
        self._proc.stdin.flush()

        deadline = time.monotonic() + timeout_s
        with self._cond:
            while req_id not in self._responses:
                if self._closed:
                    raise self._close_error or RuntimeError("MCP client is closed")
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    raise TimeoutError(f'MCP request "{method}" timed out after {timeout_s}s')
                self._cond.wait(timeout=remaining)
            msg = self._responses.pop(req_id)

        if msg.get("error"):
            err = msg["error"]
            raise RuntimeError(f"MCP server error {err.get('code')}: {err.get('message')}")
        return msg.get("result")

    def _notify(self, method: str, params: Any = None) -> None:
        if self._closed:
            return
        payload = json.dumps({"jsonrpc": "2.0", "method": method, "params": params})
        assert self._proc.stdin is not None
        self._proc.stdin.write(payload + "\n")
        self._proc.stdin.flush()

    def initialize(self) -> Any:
        result = self._request(
            "initialize",
            {
                "protocolVersion": "2024-11-05",
                "capabilities": {},
                "clientInfo": {"name": "mcpseal", "version": VERSION},
            },
        )
        self._notify("notifications/initialized")
        return result

    def list_tools(self) -> list[McpToolDefinition]:
        result = self._request("tools/list", {})
        if isinstance(result, dict):
            return result.get("tools") or []
        return []

    def close(self) -> None:
        with self._cond:
            if self._closed:
                return
            self._closed = True
            self._cond.notify_all()
        kill_process_tree(self._proc)

File: packages/cli-node/src/mcp-client.ts
// Minimal MCP stdio client: spawns a server process and speaks newline-
// delimited JSON-RPC (confirmed empirically against a real reference server
// — see Tasks.md 2.1 verification notes; the MCP stdio transport is one
// JSON-RPC message per line, no Content-Length framing).
//
// This deliberately lives in cli-node, not cli-core: cli-core is the
// canonical hashing/drift spec and CLAUDE.md says to keep it lean and
// auditable (the trust-critical path). Spawning processes and speaking a
// wire protocol is CLI/product plumbing, not part of that spec.
import { spawn, type ChildProcessWithoutNullStreams } from "node:child_process";
import { VERSION } from "./version.js";
import { killProcessTree } from "./process-utils.js";

export interface McpServerCommand {
  command: string;
  args: string[];
}

export interface McpToolDefinition {
  name: string;
  description: string;
  inputSchema: object;
  [extra: string]: unknown;
}

interface JsonRpcResponse {
  jsonrpc: "2.0";
  id?: number;
  result?: unknown;
  error?: { code: number; message: string };
}

const DEFAULT_TIMEOUT_MS = 15_000;

export class McpStdioClient {
  private child: ChildProcessWithoutNullStreams;
  private buffer = "";
  private nextId = 1;
  private pending = new Map<number, { resolve: (v: unknown) => void; reject: (e: Error) => void }>();
  private closed = false;
  private closeError: Error | undefined;

  constructor(serverCommand: McpServerCommand, options?: { cwd?: string; shell?: boolean }) {
    this.child = spawn(serverCommand.command, serverCommand.args, {
      stdio: ["pipe", "pipe", "pipe"],
      cwd: options?.cwd,
      shell: options?.shell ?? true,
    });
    // Forward the spawned process's stderr so diagnostics (e.g. mcpseal
    // proxy's own "blocked tool" logging) are visible to whoever is running
    // this client, not silently buffered in an unread pipe.
    this.child.stderr.pipe(process.stderr);

    this.child.stdout.on("data", (chunk: Buffer) => this.onStdout(chunk));
    this.child.on("error", (err) => this.onClose(err));
    this.child.on("exit", (code, signal) => {
      if (!this.closed) {
        this.onClose(new Error(`server process exited (code=${code}, signal=${signal}) before responding`));
      }
    });
  }

  private onStdout(chunk: Buffer): void {
    this.buffer += chunk.toString("utf-8");
    let idx: number;
    while ((idx = this.buffer.indexOf("\n")) !== -1) {
      const line = this.buffer.slice(0, idx);
      this.buffer = this.buffer.slice(idx + 1);
      if (!line.trim()) continue;
      this.handleLine(line);
    }
  }

  private handleLine(line: string): void {
    let msg: JsonRpcResponse;
    try {
      msg = JSON.parse(line);
    } catch {
      return; // not a JSON-RPC message we can correlate; ignore (e.g. partial/garbage)
    }
    if (msg.id === undefined) return; // notification, not a response to a request
    const waiter = this.pending.get(msg.id);
    if (!waiter) return;
    this.pending.delete(msg.id);
    if (msg.error) {
      waiter.reject(new Error(`MCP server error ${msg.error.code}: ${msg.error.message}`));
    } else {
      waiter.resolve(msg.result);
    }
  }

  private onClose(err: Error): void {
    this.closed = true;
    this.closeError = err;
    for (const [, waiter] of this.pending) {
      waiter.reject(err);
    }
    this.pending.clear();
  }

  private request(method: string, params: unknown, timeoutMs = DEFAULT_TIMEOUT_MS): Promise<unknown> {
    if (this.closed) {
      return Promise.reject(this.closeError ?? new Error("MCP client is closed"));
    }
    const id = this.nextId++;
    const payload = JSON.stringify({ jsonrpc: "2.0", id, method, params });
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(id);
        reject(new Error(`MCP request "${method}" timed out after ${timeoutMs}ms`));
      }, timeoutMs);
      this.pending.set(id, {
        resolve: (v) => {
          clearTimeout(timer);
          resolve(v);
        },
        reject: (e) => {
          clearTimeout(timer);
          reject(e);
        },
      });
      this.child.stdin.write(payload + "\n");
    });
  }

  private notify(method: string, params?: unknown): void {
    if (this.closed) return;
    const payload = JSON.stringify({ jsonrpc: "2.0", method, params });
    this.child.stdin.write(payload + "\n");
  }

  async initialize(): Promise<unknown> {
    const result = await this.request("initialize", {
      protocolVersion: "2024-11-05",
      capabilities: {},
      clientInfo: { name: "mcpseal", version: VERSION },
    });
    this.notify("notifications/initialized");
    return result;
  }

  async listTools(): Promise<McpToolDefinition[]> {
    const result = (await this.request("tools/list", {})) as { tools?: McpToolDefinition[] };
    return result.tools ?? [];
  }

  close(): void {
    if (this.closed) return;
    this.closed = true;
    killProcessTree(this.child);
  }
}
